File: src/voice/transcribe.ts
import OpenAI from "openai";
import { createReadStream } from "fs";
import { config } from "../config.js";
import { logger } from "../utils/logger.js";

let client: OpenAI | null = null;

function getClient(): OpenAI {
  if (!client) {
    if (!config.openaiApiKey) {
      throw new Error("OPENAI_API_KEY not set — voice transcription unavailable");
    }
    client = new OpenAI({ apiKey: config.openaiApiKey });
  }
  return client;
}

/**
 * Transcribe an audio file using OpenAI Whisper API.
 */
export async function transcribeAudio(filePath: string): Promise<string> {
  const openai = getClient();

  logger.debug({ filePath }, "Transcribing audio");

  const response = await openai.audio.transcriptions.create({
    model: "whisper-1",
    file: createReadStream(filePath),
    language: "en",
  });

  logger.debug(
    { text: response.text.slice(0, 100) },
    "Transcription complete",
  );

  return response.text;
}

File: proxima/voice/transcribe.py
from __future__ import annotations

from pathlib import Path

from openai import AsyncOpenAI

_client: AsyncOpenAI | None = None


def _get_client(api_key: str) -> AsyncOpenAI:
    global _client
    if _client is None:
        _client = AsyncOpenAI(api_key=api_key)
    return _client


async def transcribe_audio(
    file_path: Path, api_key: str | None, *, language: str | None = None
) -> str:
    if not api_key:
        raise RuntimeError("OPENAI_API_KEY is not configured; voice transcription is unavailable")

    client = _get_client(api_key)
    kwargs: dict[str, object] = {"model": "whisper-1"}
    if language:
        kwargs["language"] = language
    with file_path.open("rb") as file_obj:
        response = await client.audio.transcriptions.create(file=file_obj, **kwargs)  # type: ignore[arg-type]
    return response.text
